File: src/src/main/resources/mmlspark/Timer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class Timer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """


    Args:

        disableMaterialization (bool): Whether to disable timing (so that one can turn it off for evaluation) (default: true)
        logToScala (bool): Whether to output the time to the scala console (default: true)
        stage (object): The stage to time
    """

    @keyword_only
    def __init__(self, disableMaterialization=True, logToScala=True, stage=None):
        super(Timer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.Timer")
        self._cache = {}
        self.disableMaterialization = Param(self, "disableMaterialization", "disableMaterialization: Whether to disable timing (so that one can turn it off for evaluation) (default: true)")
        self._setDefault(disableMaterialization=True)
        self.logToScala = Param(self, "logToScala", "logToScala: Whether to output the time to the scala console (default: true)")
        self._setDefault(logToScala=True)
        self.stage = Param(self, "stage", "stage: The stage to time", generateTypeConverter("stage", self._cache, complexTypeConverter))
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, disableMaterialization=True, logToScala=True, stage=None):
        """
        Set the (keyword only) parameters

        Args:

            disableMaterialization (bool): Whether to disable timing (so that one can turn it off for evaluation) (default: true)
            logToScala (bool): Whether to output the time to the scala console (default: true)
            stage (object): The stage to time
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setDisableMaterialization(self, value):
        """

        Args:

            disableMaterialization (bool): Whether to disable timing (so that one can turn it off for evaluation) (default: true)

        """
        self._set(disableMaterialization=value)
        return self


    def getDisableMaterialization(self):
        """

        Returns:

            bool: Whether to disable timing (so that one can turn it off for evaluation) (default: true)
        """
        return self.getOrDefault(self.disableMaterialization)


    def setLogToScala(self, value):
        """

        Args:

            logToScala (bool): Whether to output the time to the scala console (default: true)

        """
        self._set(logToScala=value)
        return self


    def getLogToScala(self):
        """

        Returns:

            bool: Whether to output the time to the scala console (default: true)
        """
        return self.getOrDefault(self.logToScala)


    def setStage(self, value):
        """

        Args:

            stage (object): The stage to time

        """
        self._set(stage=value)
        return self


    def getStage(self):
        """

        Returns:

            object: The stage to time
        """
        return self._cache.get("stage", None)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.Timer"

    @staticmethod
    def _from_java(java_stage):
        module_name=Timer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".Timer"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return TimerModel(java_model)


class TimerModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`Timer`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.TimerModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=TimerModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".TimerModel"
        return from_java(java_stage, module_name)

File: core/src/main/scala/com/microsoft/azure/synapse/ml/stages/Timer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.stages

import com.microsoft.azure.synapse.ml.codegen.Wrappable
import com.microsoft.azure.synapse.ml.logging.{FeatureNames, SynapseMLLogging}
import com.microsoft.azure.synapse.ml.param.{PipelineStageParam, TransformerParam}
import org.apache.spark.ml._
import org.apache.spark.ml.param.{BooleanParam, ParamMap}
import org.apache.spark.ml.util._
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, Dataset}

object Timer extends ComplexParamsReadable[Timer]

trait TimerParams extends Wrappable {
  val stage = new PipelineStageParam(this, "stage", "The stage to time")

  def getStage: PipelineStage = $(stage)

  val logToScala = new BooleanParam(this, "logToScala", "Whether to output the time to the scala console")
  setDefault(logToScala -> true)

  def getLogToScala: Boolean = $(logToScala)

  def setLogToScala(v: Boolean): this.type = set(logToScala, v)

  val disableMaterialization = new BooleanParam(this, "disableMaterialization",
                 "Whether to disable timing (so that one can turn it off for evaluation)")

  setDefault(disableMaterialization->true)

  def getDisableMaterialization: Boolean = $(disableMaterialization)

  def setDisableMaterialization(v: Boolean): this.type = set(disableMaterialization, v)

  protected def formatTime(t: Long, isTransform: Boolean, count: Option[Long], stage: PipelineStage): String = {
    val time = {
      t / 1000000000.0
    }
    val action = if (isTransform) "transform" else "predict"
    val amount = count match {
      case Some(i) => s"$i rows"
      case _ => ""
    }
    s"$stage took ${time}s to $action $amount"
  }

  protected def log(str: String): Unit = {
    if (str != "") println(str)
  }

}

class Timer(val uid: String) extends Estimator[TimerModel]
  with TimerParams with ComplexParamsWritable with SynapseMLLogging {
  logClass(FeatureNames.Core)

  def this() = this(Identifiable.randomUID("Timer"))

  def setStage(s: PipelineStage): this.type = set(stage, s)

  override def copy(extra: ParamMap): this.type = defaultCopy(extra)

  override def transformSchema(schema: StructType): StructType = getStage.transformSchema(schema)

  def fitWithTime(dataset: Dataset[_]): (TimerModel, String) = {
    val cachedDatasetBefore = if (getDisableMaterialization) dataset else dataset.cache()
    val countBeforeVal = if (getDisableMaterialization) None else Some(cachedDatasetBefore.count())
    getStage match {
      case t: Transformer => (new TimerModel().setTransformer(t).setParent(this), "")
      case e: Estimator[_] =>
        val t0 = System.nanoTime()
        val fitE = e.fit(dataset)
        val t1 = System.nanoTime()
        (new TimerModel().setTransformer(fitE).setParent(this),
         formatTime(t1 - t0, false, countBeforeVal, e))
    }
  }

  def fit(dataset: Dataset[_]): TimerModel = {
    logFit({
      val (model, message) = fitWithTime(dataset)
      log(message)
      model
    }, dataset.columns.length)
  }

}

object TimerModel extends ComplexParamsReadable[TimerModel]

class TimerModel(val uid: String)
  extends Model[TimerModel] with TimerParams with ComplexParamsWritable with SynapseMLLogging {
  logClass(FeatureNames.Core)

  def this() = this(Identifiable.randomUID("TimerModel"))

  val transformer = new TransformerParam(this, "transformer", "inner model to time")

  def getTransformer: Transformer = $(transformer)

  def setTransformer(v: Transformer): this.type = set(transformer, v)

  override def copy(extra: ParamMap): TimerModel = defaultCopy(extra)

  def transformWithTime(dataset: Dataset[_]): (DataFrame, String) = {
    val cachedDatasetBefore = if (getDisableMaterialization) dataset else dataset.cache()
    val countBeforeVal = if (getDisableMaterialization) None else Some(cachedDatasetBefore.count())

    val t0 = System.nanoTime()

    val res = getTransformer.transform(cachedDatasetBefore)

    val cachedDatasetAfter = if (getDisableMaterialization) res else res.cache()
    val countAfterVal = if (getDisableMaterialization) None else Some(cachedDatasetAfter.count())

    val t1 = System.nanoTime()

    (cachedDatasetAfter, formatTime(t1 - t0, true, countAfterVal, getTransformer))
  }

  def transformSchema(schema: StructType): StructType = getTransformer.transformSchema(schema)

  def transform(dataset: Dataset[_]): DataFrame = {
    logTransform[DataFrame]({
      val (model, message) = transformWithTime(dataset)
      log(message)
      model
    }, dataset.columns.length)
  }

}
